### ISSUE TITLE
CachedNetworkImageProvider should have placeholder

### ISSUE THREAD

**Issue by feinstein** [renefloor/flutter_cached_network_image#333]
The `CachedNetworkImageProvider` should have a `placeholder` which is the first image to be sent in the `ImageProvider` `ImageStream`. This way we can easy deal with cases where an image doesn't exist on the server.

**Comment by renefloor** [comment]
@feinstein I couldn't find a way to supply multiple images to the ImageStream in an ImageProvider. If you know a way to do that please make a pull request for it. Then we can also use `cacheManager.getFile` instead of `cacheManager.getSingleFile`

**Comment by feinstein** [comment]
I am not familiar with the API, I thought there was a way for its name being `ImageStream`, but maybe it only allows one image... I will open an issue about this on the main Repo then.

**Comment by renefloor** [comment]
I think it is a stream for types like gifs that are multiple images in itself

**Comment by feinstein** [comment]
And you can't leverage this to be a placeholder? Pretending that it's a gif?

**Comment by renefloor** [comment]
Maybe, but it would not be easy I guess

**Comment by renefloor** [comment]
Soon the ImageProvider will support showing multiple images after each other, so adding a placeholder will be an option, however it won't fade.

**Comment by feinstein** [comment]
Maybe you could reach the Flutter team and propose to change the inner workings of the `ImageStream` so you could fade, like adding a callback for whenever the image changes so it could be transformed, animated etc.

**Comment by clragon** [comment]
any updates on this?
this would be perfect to show a lower res cached version of images, while the high res is loading.

**Comment by renefloor** [comment]
I didn't work on it. A placeholder in the ImageProvider should be possible to make at least, but I think the fading will be very hard or maybe even impossible to achieve.
@clragon if you use the CachedNetworkImage widget instead of the ImageProvider you can use a CachedNetworkImage as a placeholder with a url to a low res image which should be loaded faster. You could also precache the low res images so they load immediately: https://api.flutter.dev/flutter/widgets/precacheImage.html

**Comment by clragon** [comment]
Yup, thats what I'm currently doing, and it works out quite well.
Though not everything takes a widget, sometimes it would be handy to have a placeholder
when the widget requires an ImageProvider.

**Comment by slavap** [comment]
Looking at the code I see that MultiImageStreamCompleter is already used, so to add placeholder it should be minor changes? E.g. adding placeholderUrl to CachedNetworkImageProvider and extending loading methods, right?
